aleph-client/src/lib.rs

```diff
@@ -105,21 +107,66 @@ pub fn create_custom_connection<Client: FromStr + RpcClient>(
     }
 }
 
-pub fn send_xt(connection: &Connection, xt: String, xt_name: &'static str, tx_status: XtStatus) {
-    let block_hash = connection
+/// Sends transaction `xt` using `connection`. If `tx_status` is either `Finalized` or `InBlock`
+/// additionally fetches corresponding block number, which is then returned.
+/// Can panic.
+pub fn send_xt(
+    connection: &Connection,
+    xt: String,
+    xt_name: &'static str,
+    tx_status: XtStatus,
+) -> Option<BlockNumber> {
+    let hash = connection
         .send_extrinsic(xt, tx_status)
         .expect("Could not send extrinsic")
         .expect("Could not get tx hash");
-    let block_number = connection
-        .get_header::<Header>(Some(block_hash))
-        .expect("Could not fetch header")
-        .expect("Block exists; qed")
-        .number;
-    info!(
-        target: "aleph-client",
-        "Transaction {} was included in block {}.",
-        xt_name, block_number
-    );
+
+    match tx_status {
+        XtStatus::Finalized | XtStatus::InBlock => {
+            let block_number = connection
+                .get_header::<Header>(Some(hash))
+                .expect("Could not fetch header")
+                .expect("Block exists; qed")
+                .number;
+            info!(
+                target: "aleph-client",
+                "Transaction {} was included in block {}.",
+                xt_name, block_number
+            );
+            Some(block_number)
+        }
+        _ => None,
+    }
+}
+
+/// Sends transaction `xt` using `connection`. If `tx_status` is either `Finalized` or `InBlock`
+/// fetches corresponding block number, which is then returned.
+/// Recoverable.
+pub fn try_send_xt(
+    connection: &Connection,
+    xt: String,
+    xt_name: &'static str,
+    tx_status: XtStatus,
+) -> ApiResult<Option<BlockNumber>> {
```

Comment:
I would keep returning the block's hash. It's less ambiguous. In theory, there may be multiple blocks under same block height (equivocations) but block hash is unique. Usually, APIs will expect the block hash rather than height when querying for data as well.

Reply:
changed